### PR TITLE
Fix warnings revealed after enabling analyzer for projects not having access to RuntimeInformation, OSPlatofrm types

### PR DESCRIPTION
**Pull request by buyaa-n**
Fix Platform Compatibility warnings revealed after enabling analyzer for projects not having access to System.Runtime.InteropServices.RuntimeInformation, System.Runtime.InteropServices.OSPlatofrm types

Related to [Ensure that each of those projects can produce warnings/errors from the analyzer during a normal build task of #42335 ]( https://github.com/dotnet/runtime/issues/42335)

### CODE REVIEW

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs [hunk 1]
```diff
@@ -748,7 +748,7 @@ public async Task<HttpResponseMessage> SendAsyncCore(HttpRequestMessage request,
                         // Map the exception the same way as we normally do.
                         catch (Exception ex) when (MapSendException(ex, cancellationToken, out Exception mappedEx))
                         {
-                            throw mappedEx;
+                            throw;
```
Comment (bartonjs) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs:751]: This seems unrelated... and like it's probably not desired.

File: src/libraries/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.Windows.cs [hunk 2]
```diff
@@ -176,7 +176,7 @@ private static void CreateCSP(CspParameters parameters, bool randomKeyContainer,
         internal static void AcquireCsp(CspParameters cspParameters, out SafeProvHandle safeProvHandle)
         {
             Debug.Assert(cspParameters != null);
-            Debug.Assert(cspParameters.KeyContainerName == null);
+            Debug.Assert(OperatingSystem.IsWindows() && cspParameters.KeyContainerName == null);
```
Comment (bartonjs) [inline src/libraries/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.Windows.cs:179]: Separate asserts, please.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs (hunk 1 above)
Comment (buyaa-n) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs:751]: oh yeah it is caught because of a new analyzer [CA2200](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2200?view=vs-2019), but the update might be not desired

File: src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/CertificatePal.PrivateKey.cs [hunk 3]
```diff
@@ -28,10 +28,12 @@ public bool HasPrivateKey
             return GetPrivateKey<RSA>(
                 delegate (CspParameters csp)
                 {
+                    Debug.Assert(OperatingSystem.IsWindows());
```
Comment (bartonjs) [inline src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/CertificatePal.PrivateKey.cs:31]: Can we eliminate all of these asserts if we paint the whole class?  Largely I'm looking for the pattern that we want to do with new code, and all these asserts feel like they'd get old, fast.

File: src/libraries/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.Windows.cs (hunk 2 above)
Comment (buyaa-n) [reply src/libraries/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.Windows.cs:179]: The warning is caught within the Assert, `cspParameters.KeyContainerName` is windows only

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs (hunk 1 above)
Comment (bartonjs) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs:751]: Hm, that seems (to me) like a bug in that analyzer, since the thing being thrown isn't the exception variable, but a variable introduced from the exception filter call.  It's not doing the thing the analyzer says not to do ("throw ex;")

File: src/libraries/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.Windows.cs (hunk 2 above)
Comment (bartonjs) [reply src/libraries/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.Windows.cs:179]: Hm.  Alright.  I'd still rather it was separate asserts, but hopefully if someone ever violates the call contract (and calls the helper method with a null KeyContainerName) they realize "this code is in a file that says `.Windows` in the name, it's probably not the left half of this expression".

Comment (buyaa-n) [reply src/libraries/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CapiHelper.Windows.cs:179]: OK, good point

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs (hunk 1 above)
Comment (buyaa-n) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs:751]: I was not sure if it will be smart enough to map to the `mappedEx` further, most likely not, with your comment I confirmed that it only maps to the original exception `ex`, then it is definitely the analzyer bug, i guess in this case the analyzer shouldn't warn, I will file an issue for the analzyer

File: src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/CertificatePal.PrivateKey.cs (hunk 3 above)
Comment (buyaa-n) [reply src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/CertificatePal.PrivateKey.cs:31]: Byt the design MSBuild emitting `SupportedOSPlatform("windows")` attribute for the builds targeting windows, which is working well for normal projects and the analyzer wouldn't warn, not sure if/how that is/could applicable for the runtime projects. But it would be great if we could not warn for windows only builds in runtime

Comment (marek-safar) [reply src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/CertificatePal.PrivateKey.cs:31]: Why is it needed? Could we make Pal.Windows to be included in Windows builds only?

Comment (bartonjs) [reply src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/CertificatePal.PrivateKey.cs:31]: > Could we make Pal.Windows to be included in Windows builds only?

It is...
https://github.com/dotnet/runtime/blob/3be4a5719a3a78c2a1e60c928c10aa48f2da69fd/src/libraries/System.Security.Cryptography.X509Certificates/src/System.Security.Cryptography.X509Certificates.csproj#L231-L234

The only place I notice that we apply the OS-specificness to an assembly is https://github.com/dotnet/runtime/blob/196fd1804dcc36fc6a398bcc318e07a171a0fc9c/eng/versioning.targets#L25-L30, and the only assemblies that do that are always Windows-only.  Seems like we have a gap for our "Windows-specific implementation assembly" assemblies.

Comment (buyaa-n) [reply src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/CertificatePal.PrivateKey.cs:31]: From what i am seeing the windows targets of runtime projects are not the same as the standard one `net5.0-windows` (has suffix `_NT` like `<TargetFrameworks>$(NetCoreAppCurrent)-Windows_NT;</TargetFrameworks>`). That means MSBuild is not emitting the desired attribute. But i am not sure even fixing the suffix would fix the issue, do not know well how the runtime build works cc @safern @ViktorHofer

Comment (buyaa-n) [reply src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/CertificatePal.PrivateKey.cs:31]: > Could we make Pal.Windows to be included in Windows builds only?

Even it is included in Windows build it would  still warn

Comment (jeffhandley) [reply src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/CertificatePal.PrivateKey.cs:31]: Repro'ing in another project, using a tfm of `net5.0-windows` yields the expected behavior of not producing warnings for calling Windows-only methods.

> That means MSBuild is not emitting the desired attribute

@buyaa-n I forget which attribute it is that gets emitted when `net5.0-windows` is used, and where that is picked up. Can you remind me what property gets set by that and how the analyzer receives it and uses it?

Comment (buyaa-n) [reply src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/CertificatePal.PrivateKey.cs:31]: > Repro'ing in another project, using a tfm of net5.0-windows yields the expected behavior of not producing warnings for calling Windows-only methods.

Could you try with `net5.0-windows7.0`? as there is a [bug ](https://github.com/dotnet/sdk/issues/13494)that MSBuild expecting a version (which is fixed but might not deployed yet

>  I forget which attribute it is that gets emitted when net5.0-windows is used, and where that is picked up. Can you remind me what property gets set by that and how the analyzer receives it and uses it?

It should emit [SupportedOSPlatform("windows")] attribute in AssemblyInfo.cs file, which will have an assembly-level application of SupportedOSPlatformAttribute.

Comment (buyaa-n) [reply src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/CertificatePal.PrivateKey.cs:31]: just `net5.0-windows` without a version is working for me, i have installed 5.0 sdk RC2 a several days ago

Comment (buyaa-n) [reply src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/CertificatePal.PrivateKey.cs:31]: > The only place I notice that we apply the OS-specificness to an assembly is runtime/eng/versioning.targets
> 
>  , and the only assemblies that do that are always Windows-only. Seems like we have a gap for our "Windows-specific implementation assembly" assemblies.

Seems i can add there [SupportedOSPlatform("windows")] attribute for windows specific builds, thanks @bartonjs, it is solved at least libraries warnings, but not corelib. I am raising another PR for fix https://github.com/dotnet/runtime/pull/43097 and closing this draft

**Comment by buyaa-n** [comment]
Raising another PR for fix #43097
